frontends/common/resolveReferences/resolveReferences.cpp

```diff
@@ -19,112 +19,106 @@ limitations under the License.
-            switch (type) {
-                case P4::ResolutionType::Any:
-                    break;
-                case P4::ResolutionType::Type: {
-                    std::function<bool(const IR::IDeclaration*)> kindFilter =
-                            [](const IR::IDeclaration* d) {
-                        return d->is<IR::Type>();
-                    };
-                    decls = decls->where(kindFilter);
-                    break;
-                }
-                case P4::ResolutionType::TypeVariable: {
-                    std::function<bool(const IR::IDeclaration*)> kindFilter =
-                            [](const IR::IDeclaration* d) {
-                    return d->is<IR::Type_Var>(); };
-                    decls = decls->where(kindFilter);
-                    break;
-                }
-            default:
-                BUG("Unexpected enumeration value %1%", static_cast<int>(type));
-            }
-
-            if (!forwardOK) {
-                std::function<bool(const IR::IDeclaration*)> locationFilter =
-                        [name](const IR::IDeclaration* d) {
-                    Util::SourceInfo nsi = name.srcInfo;
-                    Util::SourceInfo dsi = d->getNode()->srcInfo;
-                    bool before = dsi <= nsi;
-                    LOG2("\tPosition test:" << dsi << "<=" << nsi << "=" << before);
-                    return before;
-                };
-                decls = decls->where(locationFilter);
-            }
-
-            auto vector = decls->toVector();
-            if (!vector->empty()) {
-                LOG2("Resolved in " << dbp(current->getNode()));
-                return vector;
-            } else {
-                continue;
-            }
-        } else {
-            auto simple = current->to<IR::ISimpleNamespace>();
-            auto decl = simple->getDeclByName(name);
-            if (decl == nullptr)
-                continue;
-            switch (type) {
-                case P4::ResolutionType::Any:
-                    break;
-                case P4::ResolutionType::Type: {
-                    if (!decl->is<IR::Type>())
-                        continue;
-                    break;
-                }
-                case P4::ResolutionType::TypeVariable: {
-                    if (!decl->is<IR::Type_Var>())
-                        continue;
-                    break;
-                }
-            default:
-                BUG("Unexpected enumeration value %1%", static_cast<int>(type));
-            }
-
-            if (!forwardOK) {
+    std::vector<const IR::IDeclaration*> *rv = &empty;
+
+    // FIXME -- we look up 'globals' solely to find match kinds (that the only thing
+    // in globals), which menas that match kinds are effectively reserved words.  Should
+    // probably move resolving match kinds out of here altogether
+    for (auto it = globals.begin(); rv->empty() && it != globals.end(); ++it)
```

Comment:
In the other implementation the globals are tried only at the very end.
This seems to try them at the beginning.

Reply:
The old code appends the globals to the end of the `toTry` vector then iterates over `toTry` in *reverse*, so the globals are checked first, not last.  That seems wrong to me, but for now I was just attempting to duplicate the former behavior without change.